### ISSUE TITLE
TES should gracefully exit when Kubernetes sends SIGTERM

### ISSUE THREAD

**Issue by MattMcL4475** [microsoft/ga4gh-tes#377]
When a Kubernetes cluster needs to restart or terminate a pod, it follows a graceful termination process. For a C# application running in a container within a Kubernetes pod, the entire process can be outlined as follows:

1. **Kubernetes Sends SIGTERM**: When Kubernetes decides to stop a pod, it first sends a `SIGTERM` signal to the root process (PID 1) in each container in the pod.

2. **C# `ProcessExit` Event Fires**: Upon receiving the `SIGTERM` signal, the .NET runtime invokes any registered event handlers for the `AppDomain.CurrentDomain.ProcessExit` event in the C# application. Your code can perform any necessary clean-up during this time, such as flushing logs, closing database connections, or shutting down services gracefully.

    ```csharp
    AppDomain.CurrentDomain.ProcessExit += new EventHandler(ProcessExit);
    ```

    Inside `ProcessExit`:

    ```csharp
    static void ProcessExit(object sender, EventArgs e)
    {
        Console.WriteLine("SIGTERM received, performing clean-up...");
        // Perform clean-up here
    }
    ```

3. **Grace Period**: The container has a set period to perform its shut-down activities, defined by the `terminationGracePeriodSeconds` parameter in the pod spec. By default, this is 30 seconds.

4. **Kubernetes Checks Container Exit Status**: Kubernetes will check if the container has exited gracefully (i.e., the process terminated on its own) within the `terminationGracePeriodSeconds`.

5. **SIGKILL Sent (if needed)**: If the container has not exited within the grace period, Kubernetes forcibly terminates the container by sending a `SIGKILL` signal. This will forcefully shut down the container, and no further clean-up can be performed at this stage.

6. **Pod Deletion**: Once all containers in the pod have either terminated gracefully or been killed, Kubernetes proceeds to delete the pod.

7. **Node Cleanup**: Any associated resources with the pod, like volumes, are also cleaned up as necessary.

In summary, Kubernetes sends a `SIGTERM` initially, and your C# code can catch this via the `ProcessExit` event. If the container doesn't exit within the time specified by `terminationGracePeriodSeconds`, then Kubernetes sends a `SIGKILL` to forcibly terminate the container.
